### PR TITLE
[VL] Support missing types for struct literal

### PR DESCRIPTION
**Pull request by zhli1142015**
## What changes were proposed in this pull request?

There is some fallback which is triggered by literal not support errors in Velox substrait, for example below one. This pull request addresses the mentioned error and includes support for other missing types for struct literals.
`E20231229 02:32:26.550776  2535 Exceptions.h:69] Line: /__w/1/s/Gluten/cpp/velox/substrait/SubstraitToVeloxExpr.cc:533, Function:literalsToRowVector, Expression:  literalsToRowVector not supported for type case '7', Source: RUNTIME, ErrorCode: NOT_IMPLEMENTED`

## How was this patch tested?
UT.

### CODE REVIEW

**Comment by zhli1142015** [comment]
Hello @JkSelf  and @rui-mo ,
Could you help to review this PR?
I want to solve this `Function:literalsToVector, Expression:  literals not supported for type case 'XX'` by this pr. But after this chagne I observed some UT failures. Looks these UT fails just because they offload to velox with this change.

```
2024-01-04T02:48:13.1317060Z [31m- typeCoercion/native/stringCastAndExpressions.sql *** FAILED ***[0m
2024-01-04T02:48:13.1319995Z [31m  typeCoercion/native/stringCastAndExpressions.sql[0m
2024-01-04T02:48:13.1322724Z [31m  Expected "{"a":[null]}", but got "{"a":[0]}" Result did not match for query #31[0m
2024-01-04T02:48:13.1325730Z [31m  select from_json(a, 'a INT') from t (GlutenSQLQueryTestSuite.scala:580)[0m
```
```
SPARK-27671: Fix analysis exception when casting null in nested field in struct *** FAILED ***[0m
2024-01-04T04:13:01.5647584Z [31m  !== Correct Answer - 1 ==   == Gluten Answer - 1 ==[0m
2024-01-04T04:13:01.5649896Z [31m  !struct<>                   struct<struct(1, NULL):struct<a:int,b:int>>[0m
2024-01-04T04:13:01.5652360Z [31m  ![[1,null]]                 [[1,0]] (GlutenSQLTestsTrait.scala:141)[0m
```
These first two may relate to null handling in `VeloxColumnarToRowExec`. I disabled them, may check them later.
image

Spark 34
This one is related to paruet writer, I don't have much context for it. I disable native writer for `GlutenVectorizedParquetReadSchemaSuite`.
```
2024-01-04T04:09:27.6290355Z Caused by: java.lang.RuntimeException: Exception: VeloxUserError
2024-01-04T04:09:27.6291084Z Error Source: USER
2024-01-04T04:09:27.6291513Z Error Code: INVALID_ARGUMENT
2024-01-04T04:09:27.6292044Z Reason: Field name must not be empty.
2024-01-04T04:09:27.6292619Z Retriable: False
2024-01-04T04:09:27.6293125Z Expression: !name.empty()
2024-01-04T04:09:27.6294098Z Function: validateSchemaRecursive
2024-01-04T04:09:27.6294987Z File: ../../velox/dwio/parquet/writer/Writer.cpp
2024-01-04T04:09:27.6296084Z Line: 151
2024-01-04T04:09:27.6296633Z Stack trace:
2024-01-04T04:09:27.6300286Z # 0  facebook::velox::VeloxException::VeloxException(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, facebook::velox::VeloxException::Type, std::basic_string_view<char, std::char_traits<char> >)
2024-01-04T04:09:27.6304409Z # 1  void facebook::velox::detail::veloxCheckFail<facebook::velox::VeloxUserError, char const*>(facebook::velox::detail::VeloxCheckFailArgs const&, char const*)
2024-01-04T04:09:27.6308255Z # 2  facebook::velox::parquet::(anonymous namespace)::validateSchemaRecursive(std::shared_ptr<facebook::velox::RowType const> const&)
2024-01-04T04:09:27.6310614Z # 3  facebook::velox::parquet::(anonymous namespace)::validateSchemaRecursive(std::shared_ptr<facebook::velox::RowType const> const&)
2024-01-04T04:09:27.6313662Z # 4  facebook::velox::parquet::Writer::Writer(std::unique_ptr<facebook::velox::dwio::common::FileSink, std::default_delete<facebook::velox::dwio::common::FileSink> >, facebook::velox::parquet::WriterOptions const&, std::shared_ptr<facebook::velox::memory::MemoryPool>, std::shared_ptr<facebook::velox::RowType const>)
2024-01-04T04:09:27.6320189Z # 5  facebook::velox::parquet::Writer::Writer(std::unique_ptr<facebook::velox::dwio::common::FileSink, std::default_delete<facebook::velox::dwio::common::FileSink> >, facebook::velox::parquet::WriterOptions const&, std::shared_ptr<facebook::velox::RowType const>)
2024-01-04T04:09:27.6323543Z # 6  facebook::velox::parquet::ParquetWriterFactory::createWriter(std::unique_ptr<facebook::velox::dwio::common::FileSink, std::default_delete<facebook::velox::dwio::common::FileSink> >, facebook::velox::dwio::common::WriterOptions const&)
2024-01-04T04:09:27.6327047Z # 7  facebook::velox::connector::hive::HiveDataSink::appendWriter(facebook::velox::connector::hive::HiveWriterId const&)
2024-01-04T04:09:27.6329982Z # 8  facebook::velox::connector::hive::HiveDataSink::appendData(std::shared_ptr<facebook::velox::RowVector>)
2024-01-04T04:09:27.6332145Z # 9  facebook::velox::exec::TableWriter::addInput(std::shared_ptr<facebook::velox::RowVector>)
2024-01-04T04:09:27.6334094Z # 10 facebook::velox::exec::Driver::runInternal(std::shared_ptr<facebook::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, std::shared_ptr<facebook::velox::RowVector>&)
2024-01-04T04:09:27.6336067Z # 11 facebook::velox::exec::Driver::next(std::shared_ptr<facebook::velox::exec::BlockingState>&)
2024-01-04T04:09:27.6337317Z # 12 facebook::velox::exec::Task::next(folly::SemiFuture<folly::Unit>*)
2024-01-04T04:09:27.6338515Z # 13 gluten::WholeStageResultIterator::next()
2024-01-04T04:09:27.6340002Z # 14 Java_io_glutenproject_vectorized_ColumnarBatchOutIterator_nativeHasNext
2024-01-04T04:09:27.6341525Z # 15 0x00007fc6bd2886c0
2024-01-04T04:09:27.6342404Z 
2024-01-04T04:09:27.6343686Z 	at io.glutenproject.vectorized.ColumnarBatchOutIterator.nativeHasNext(Native Method)
2024-01-04T04:09:27.6345373Z 	at io.glutenproject.vectorized.ColumnarBatchOutIterator.hasNextInternal(ColumnarBatchOutIterator.java:65)
2024-01-04T04:09:27.6347176Z 	at io.glutenproject.vectorized.GeneralOutIterator.hasNext(GeneralOutIterator.java:37)
2024-01-04T04:09:27.6348181Z 	... 19 more
2024-01-04T04:09:27.6349298Z [31m- add a nested column at the end of the leaf struct column *** FAILED ***[0m
```

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc [hunk 1]
```diff
@@ -519,19 +519,98 @@ RowVectorPtr SubstraitVeloxExprConverter::literalsToRowVector(const ::substrait:
...
+      }
+      case ::substrait::Expression_Literal::LiteralTypeCase::kDecimal: {
+        auto decimal = child.decimal().value();
+        auto precision = child.decimal().precision();
+        auto scale = child.decimal().scale();
+        int128_t decimalValue;
+        memcpy(&decimalValue, decimal.c_str(), 16);
+        auto type = DECIMAL(precision, scale);
+        if (precision <= 18) {
+          auto vector = BaseVector::create(type, 1, pool_);
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxExpr.cc:584]: Move to above the if clause?

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc [hunk 2]
```diff
@@ -519,19 +519,98 @@ RowVectorPtr SubstraitVeloxExprConverter::literalsToRowVector(const ::substrait:
...
   }
-  auto typeCase = structLiteral.struct_().fields(0).literal_type_case();
-  switch (typeCase) {
-    case ::substrait::Expression_Literal::LiteralTypeCase::kBinary: {
-      std::vector<VectorPtr> vectors;
-      vectors.reserve(structLiteral.struct_().fields().size());
-      for (auto& child : structLiteral.struct_().fields()) {
+  std::vector<VectorPtr> vectors;
+  vectors.reserve(structLiteral.struct_().fields().size());
+  for (auto& child : structLiteral.struct_().fields()) {
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxExpr.cc:524]: Not your change. Better to use `const &`.

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc [hunk 3]
```diff
@@ -519,19 +519,98 @@ RowVectorPtr SubstraitVeloxExprConverter::literalsToRowVector(const ::substrait:
...
+          auto vector = BaseVector::create(type, 1, pool_);
+          auto flatVector = vector->as<FlatVector<int64_t>>();
+          flatVector->set(0, static_cast<int64_t>(decimalValue));
+          vectors.emplace_back(vector);
+        } else {
+          auto vector = BaseVector::create(type, 1, pool_);
+          auto flatVector = vector->as<FlatVector<int128_t>>();
+          flatVector->set(
+              0, HugeInt::build(static_cast<uint64_t>(decimalValue >> 64), static_cast<uint64_t>(decimalValue)));
+          vectors.emplace_back(vector);
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxExpr.cc:593]: Could be moved out from the if-else clause.

File: gluten-ut/spark33/src/test/scala/io/glutenproject/utils/velox/VeloxSQLQueryTestSettings.scala [hunk 4]
```diff
@@ -226,7 +226,8 @@ object VeloxSQLQueryTestSettings extends SQLQueryTestSettings {
     "typeCoercion/native/mapconcat.sql",
     "typeCoercion/native/mapconcat.sql",
     "typeCoercion/native/promoteStrings.sql",
-    "typeCoercion/native/stringCastAndExpressions.sql",
+    // Result different: Expected "{"a":[null]}", but got "{"a":[0]}"
+    // "typeCoercion/native/stringCastAndExpressions.sql",
```
Comment (rui-mo) [inline gluten-ut/spark33/src/test/scala/io/glutenproject/utils/velox/VeloxSQLQueryTestSettings.scala]: Could you help open an issue for this bug and put the link here?

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    validateOffloadResult("SELECT struct(-0.1, array(5, 6))")
+    validateOffloadResult("SELECT struct(7, map('red', 1, 'green', 2))")
+    validateOffloadResult("SELECT struct(array(5, 6), map('red', 1, 'green', 2))")
+    validateOffloadResult("SELECT struct(1.0, struct(array(5, 6), map('red', 1)))")
+    validateOffloadResult("SELECT struct(5, 1S, 1Y, -1Y, true, false)")
+    validateOffloadResult("SELECT struct(1D, 1F)")
+    validateOffloadResult("SELECT struct(5.321E2BD, 0.1, 5.321E22BD)")
+    validateOffloadResult("SELECT struct(TIMESTAMP'2020-12-31')")
+    validateOffloadResult("SELECT struct(X'1234')")
+    validateOffloadResult("SELECT struct(DATE'2020-12-31')")
```
Comment (rui-mo) [inline backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala:63]: Could you also add test for decimal and IntervalDayToSecond if they are to be supported?

**Review by rui-mo — COMMENTED** [review]
Thanks for your fix.

File: gluten-ut/spark33/src/test/scala/io/glutenproject/utils/velox/VeloxSQLQueryTestSettings.scala (hunk 4 above)
Comment (zhli1142015) [reply gluten-ut/spark33/src/test/scala/io/glutenproject/utils/velox/VeloxSQLQueryTestSettings.scala]: `Null` is not handled correctly for struct literal, adding below should fix this UT.
https://github.com/oap-project/gluten/pull/4267/files#diff-66c2133a2e9ec8f7a3c631d4aebb91e9e6e081f667b628e43e1754391aa2ec38R33-R35

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala (hunk 5 above)
Comment (zhli1142015) [reply backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala:63]: Decimal is tested in Line 60. For  IntervalDayToSecond , looks this type is not supported in core module yet: https://github.com/oap-project/gluten/blob/c47112f6a5415ce13b43348e0453513da043fcff/gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala#L398

```
scala> spark.sql("SELECT struct(CAST('11 23:4:0' AS INTERVAL DAY TO SECOND))").collect
24/01/04 16:19:23 WARN ProjectExecTransformer: Validation failed with exception for plan: ProjectExecTransformer, due to:
24/01/04 16:19:23 WARN GlutenFallbackReporter: Validation failed for plan: Project[QueryId=0], due to: Type DayTimeIntervalType(0,3) not supported..
res0: Array[org.apache.spark.sql.Row] = Array([[PT287H4M]])
```

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc (hunk 2 above)
Comment (zhli1142015) [reply cpp/velox/substrait/SubstraitToVeloxExpr.cc]: Updated, thanks.

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc (hunk 1 above)
Comment (zhli1142015) [reply cpp/velox/substrait/SubstraitToVeloxExpr.cc]: Updated, thanks.

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.execution
+
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.execution.ProjectExec
+
+class VeloxLiteralSuite extends VeloxWholeStageTransformerSuite {
+  override protected val resourcePath: String = "/tpch-data-literal-velox"
```
Comment (JkSelf) [inline backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala:23]: Does this resourcePath exist?

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala (hunk 5 above)
Comment (JkSelf) [reply backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala:63]: Do we need to add nested type check, map(struct), array(struct) ?

File: gluten-ut/spark34/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala [hunk 7]
```diff
@@ -943,6 +943,8 @@ class VeloxTestSettings extends BackendTestSettings {
     .exclude("SPARK-39557 INSERT INTO statements with tables with array defaults")
     .exclude("SPARK-39557 INSERT INTO statements with tables with struct defaults")
     .exclude("SPARK-39557 INSERT INTO statements with tables with map defaults")
+    // Field name must not be empty.
```
Comment (JkSelf) [inline gluten-ut/spark34/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:946]: Can you record the disable unit tests in https://github.com/oap-project/gluten/issues/3559? And I will pick them after this PR is merged. Thanks.

**Review by JkSelf — COMMENTED** [review]
@zhli1142015 Thanks for your work. LGTM except a few questions.

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala (hunk 5 above)
Comment (zhli1142015) [reply backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala:63]: These should not be supported yet, I noticed it in our log also. will work on it later.
Function:literalsToVector, Expression:  literals not supported for type case '25', Source: RUNTIME, ErrorCode: NOT_IMPLEMENTED

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala (hunk 6 above)
Comment (zhli1142015) [reply backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala:23]: By now it's not used, just a placeholder.

File: gluten-ut/spark34/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala (hunk 7 above)
Comment (zhli1142015) [reply gluten-ut/spark34/src/test/scala/io/glutenproject/utils/velox/VeloxTestSettings.scala:946]: sure, I disable native writer in *ParquetReadSchemaSuite because of same error, FYI.

**Review by JkSelf — APPROVED** [review]
LGTM. Thanks.

File: gluten-core/src/main/java/io/glutenproject/substrait/expression/StructLiteralNode.java [hunk 8]
```diff
@@ -30,6 +30,9 @@ public StructLiteralNode(InternalRow row, TypeNode typeNode) {
   public LiteralNode getFieldLiteral(int index) {
     InternalRow value = getValue();
     TypeNode type = ((StructNode) getTypeNode()).getFieldTypes().get(index);
+    if (value.isNullAt(index)) {
+      return ExpressionBuilder.makeNullLiteral(type);
```
Comment (ulysses-you) [inline gluten-core/src/main/java/io/glutenproject/substrait/expression/StructLiteralNode.java:34]: can we add a test case for null value in struct literal ?

Comment (zhli1142015) [reply gluten-core/src/main/java/io/glutenproject/substrait/expression/StructLiteralNode.java:34]: This change here makes sure null value is not converted to unexpected value, for example 0 for int. Now we still not support null in side struct literal. there are many UT fail with it. Will fix and enable it later.
Thanks.

**Review by ulysses-you — APPROVED** [review]
